pkg/protocol/engines.go

```diff
@@ -0,0 +1,261 @@
+	"github.com/iotaledger/hive.go/runtime/module"
+	"github.com/iotaledger/hive.go/runtime/workerpool"
+	"github.com/iotaledger/iota-core/pkg/protocol/engine"
+	"github.com/iotaledger/iota-core/pkg/protocol/engine/accounts/accountsledger"
+	"github.com/iotaledger/iota-core/pkg/protocol/engine/blocks"
+	"github.com/iotaledger/iota-core/pkg/protocol/engine/eviction"
+	"github.com/iotaledger/iota-core/pkg/storage"
+	"github.com/iotaledger/iota-core/pkg/storage/utils"
+	iotago "github.com/iotaledger/iota.go/v4"
+)
+
+// Engines is a subcomponent of the protocol that exposes the engines that are managed by the protocol.
+type Engines struct {
+	// Main contains the main engine.
+	Main reactive.Variable[*engine.Engine]
+
+	// protocol contains a reference to the Protocol instance that this component belongs to.
+	protocol *Protocol
+
+	// worker contains the worker pool that is used to process changes to the engine instances asynchronously.
+	worker *workerpool.WorkerPool
+
+	// directory contains the directory that is used to store the engine instances on disk.
+	directory *utils.Directory
+
+	// ReactiveModule embeds a reactive module that provides default API for logging and lifecycle management.
+	*module.ReactiveModule
+}
+
+// newEngines creates a new Engines instance.
+func newEngines(protocol *Protocol) *Engines {
+	e := &Engines{
+		Main:           reactive.NewVariable[*engine.Engine](),
+		ReactiveModule: protocol.NewReactiveSubModule("Engines"),
+		protocol:       protocol,
+		worker:         protocol.Workers.CreatePool("Engines", workerpool.WithWorkerCount(1)),
+		directory:      utils.NewDirectory(protocol.Options.BaseDirectory),
+	}
+
+	protocol.Constructed.OnTrigger(func() {
+		shutdown := lo.Batch(
+			e.syncMainEngineFromMainChain(),
+			e.syncMainEngineInfoFile(),
+			e.injectEngineInstances(),
+		)
+
+		e.Shutdown.OnTrigger(func() {
+			shutdown()
+
+			e.Stopped.Trigger()
+		})
+
+		e.Initialized.Trigger()
+	})
+
+	e.Constructed.Trigger()
+
+	return e
+}
+
+// LoadMainEngine loads the main engine from disk or creates a new one if no engine exists.
+func (e *Engines) LoadMainEngine(snapshotPath string) (*engine.Engine, error) {
+	info := &engineInfo{}
+	if err := ioutils.ReadJSONFromFile(e.infoFilePath(), info); err != nil && !ierrors.Is(err, os.ErrNotExist) {
+		return nil, ierrors.Errorf("unable to read engine info file: %w", err)
+	}
+
+	// load previous engine as main engine if it exists.
+	if len(info.Name) > 0 {
+		if exists, isDirectory, err := ioutils.PathExists(e.directory.Path(info.Name)); err == nil && exists && isDirectory {
+			e.Main.Set(e.loadEngineInstanceFromSnapshot(info.Name, snapshotPath))
+		}
+	}
+
+	// load new engine if no previous engine exists.
+	e.Main.Compute(func(mainEngine *engine.Engine) *engine.Engine {
+		if mainEngine != nil {
+			return mainEngine
+		}
+
+		return e.loadEngineInstanceFromSnapshot(lo.PanicOnErr(uuid.NewUUID()).String(), snapshotPath)
+	})
```

Comment:
Why isn't this entire block part of `e.Main.Compute`? Above you use set and then compute and even check the value again if it is nil. Why is that?

Reply:
dunno, I changed it.